### DISCUSSION TITLE
Better Xcloud APK

### DISCUSSION THREAD

**Discussion by Myst3XFS** [redphx/better-xcloud#282] [General]
Is there the possibility of creating an app for Android TV / Mobile (webview), which runs xcloud directly with better xcloud injected, and which has a remap for controls that don't work correctly like dual sense and others?

**Comment by redphx** [comment]
I'd love to, but there are a few reasons why I didn't do it (yet):  

- Android WebView is similar to Chromium, but it doesn't have all of Chromium's features (Main-profile codec for example) so you'll have a worse experience.
- It adds more work for me.

**Comment by KingNothing81** [comment]
@Myst3XFS : In another topic, user @keanew already commented on August, 2023 ...

> I don't really know. I'm trying to integrate "Better xcloud" in a webview based android APK. So I don't really know what the browser the smart tv or the tablet is using. But I think this is a "chromium" browser integrate in the webview.
>
>No worry for usage of your code ^^. I'm just testing out this solution to verify if performances are better or not on an "native" apk.
And the smart TV can't install Kiwi browser. So it could be a solution.

https://github.com/redphx/better-xcloud/discussions/113#discussioncomment-6827916
 
So obviously the user already tried making such an APK for personal testing. I left a comment saying that this would be a great solution for Android TV users, but unfortunately I never got a response. It would have been interesting to know how such an APK performed and if the user ended up using it or not.

> Android WebView is similar to Chromium, but it doesn't have all of Chromium's features (Main-profile codec for example) so you'll have a worse experience.

@redphx: Are you sure about that? Trying to find a better solution than Kiwi Browser, I came across "XBrowser". 

https://play.google.com/store/apps/details?id=com.xbrowser.play
https://en.xbext.com/index.html

It's a small, minimalistic browser. The APK is just about 1 MB in size. So I think it must use the system's WebView. It even supports installing Tampermonkey scripts directly. I tried installing "Better xCloud", but it had some glitches. For example showing an error message for a short moment everytime I started a game. Or sometimes the script's settings showed up in gigantic size. But according to the badges in the stream menu, I was getting the high codec profile. And judging by my eyes, the stream quality seemed as good as with Kiwi. But of course the badges could show the wrong info. And I don't know if I could really see the difference between the medium (base) and high (main) codec profile.

**Reply by Myst3XFS** [reply]
I tried to add the script without tampermonkey in the webview, however I had some problems when trying to simulate @Run-at document-start.

Some bugs occurred despite the script partially running, the menu appeared and the options were easily modifiable, but the regions did not load, I believe this occurred because the injection is being performed at the wrong time in the page's life cycle.

![regions](https://github.com/redphx/better-xcloud/assets/162530207/b20ac309-0165-4a74-9092-328a10e2dce9)

**Reply by Myst3XFS** [reply]
I didn't know xbrowser, i find it interesting, i'll analyze it!

**Reply by redphx** [reply]
There are many WebView/WebKit browers claim to have Userscript support but they simply inject the JS files into the page. For Better xCloud to work it has to be injected before anything else (#run-at document-start). If you see the ??? button it means the script wasn't injected properly.  

Recently Android has this new API should fix this problem, but it requires at least Android 14, which is mostly useless: https://developer.android.com/reference/androidx/webkit/WebViewCompat#addDocumentStartJavaScript(android.webkit.WebView,java.lang.String,java.util.Set%3Cjava.lang.String%3E)

**Reply by KingNothing81** [reply]
Yes, I think the glitches I see when using the scipt in xBrowser are also caused by the script being auto injected at the wrong time. The advantage of using xBrowser would have been that it has a real fullscreen mode that also works on Android TV and hides all UI elements. And that keeps enabled even when the app was closed. And it opens and loads the xCloud page so much faster. Kiwi seems to always take a nap before loading the webpage after it was closed and you open it again. But I will just stick to using Kiwi Browser.

By the way ... Kiwi Browser also has a GitHub page. Where they release more frequent updates than on the PlayStore. So if someone has to sideload it anyway, you may want to think about getting it from there.

https://github.com/kiwibrowser/src.next/releases

**Reply by KingNothing81** [reply]
@Myst3XFS: Maybe also interesting: There used to be a stand-alone Tampermonkey app for Android.
https://play.google.com/store/apps/details?id=net.biniok.tampermonkey

But it hasn't been updated for many years. But I guess it was more or less just a web browser using WebView with Tampermonkey implemented. So maybe that could be an approch of creating a "Better xCloud" app?

**Reply by Myst3XFS** [reply]
It's a possibility, we'll only find out by trying, I'll check how it works and I'll try to replicate it to test the behavior.

**Reply by Myst3XFS** [reply]
@redphx 
I haven't had time to study Tampermonkey recommended by @kingNothing81 yet, nor the Xbrowser you recommended. That's because I'm focused on work projects during business hours. The little time I had left in the week I tried to play around a bit with the webview. I managed to use some things from the script directly in the webview, but the interface seems larger than it should be.

![image](https://github.com/redphx/better-xcloud/assets/162530207/badc9d20-9070-4001-89bc-25f62bfdb267)
![image](https://github.com/redphx/better-xcloud/assets/162530207/a31aa031-cd75-43bf-9481-c969c31fb567)
![image](https://github.com/redphx/better-xcloud/assets/162530207/34e892d8-97dd-4af8-904a-acceac1a54af)
![image](https://github.com/redphx/better-xcloud/assets/162530207/73a1fd85-2716-4e80-bb67-33623ce870cb)

**Reply by redphx** [reply]
That looks promising. Is it running Android 14?  

Have your tried this to fix the zoom-in problem: https://stackoverflow.com/questions/5448841/what-do-setusewideviewport-and-setloadwithoverviewmode-precisely-do

**Reply by Myst3XFS** [reply]
I will try now.
" https://stackoverflow.com/questions/5448841/what-do-setusewideviewport-and-setloadwithoverviewmode-precisely-do"

**Reply by Myst3XFS** [reply]
@redphx looks better now
![image](https://github.com/redphx/better-xcloud/assets/162530207/03356508-f5f9-4294-8b68-1b4b5c2e9350)
![image](https://github.com/redphx/better-xcloud/assets/162530207/67bf4069-96f2-4db6-b3c4-994a4053e634)

**Reply by redphx** [reply]
Thanks for testing. I was able to create an app myself too. I'll try to make a quick app this weekend.

![2024-03-13 11 29 13](https://github.com/redphx/better-xcloud/assets/96280/ba44442f-7150-4ba2-b998-bd61ee76d3a8)

**Reply by Myst3XFS** [reply]
@redphx nice bro.

**Comment by KingNothing81** [comment]
@Myst3XFS: By the way ... have you seen this? 
https://github.com/redphx/better-xcloud/discussions/234

If you are using "Better xCloud" on Android TV, you may find it useful. It's not what you want. But at least it will give you a shortcut for your Android TV home screen. So it feels a little bit like having a "Better xCloud" app. ;)

**Reply by Myst3XFS** [reply]
@KingNothing81 Interesting, I will use it.

**Comment by ewangke** [comment]
@redphx Is it possible to integrate Better xCloud on iOS with WKWebView? I'm glad to creating an iOS open source app if it's possible.

**Reply by redphx** [reply]
Sorry I don't know anything about iOS.  

It's only possible if it has an API similar to this on Android: https://developer.android.com/reference/androidx/webkit/WebViewCompat#addDocumentStartJavaScript(android.webkit.WebView,java.lang.String,java.util.Set%3Cjava.lang.String%3E)

**Reply by ewangke** [reply]
It looks promising, I'll try the iOS counterpart later. Many thanks!

**Reply by Myst3XFS** [reply]
https://github.com/react-native-webview/react-native-webview/blob/master/docs/Reference.md
This link has a method called injectedJavaScriptBeforeContentLoaded
In its description there is information about what is used on IOS and Android

**Comment by redphx** [comment]
@Myst3XFS @KingNothing81 the Android app is now available for testing: https://better-xcloud.github.io/android/

**Reply by Myst3XFS** [reply]
I have a question, will the app only work on Android 14+?

**Reply by redphx** [reply]
It supports Android 8+

**Reply by Myst3XFS** [reply]
When I try to use Android 8, 9, 10, the Xbox website exhibits strange behaviors, such as the profile bar not triggering the login button popup and other screen elements seeming unresponsive when clicked. I must have done something wrong, probably. I'll create new devices and test again.

**Reply by redphx** [reply]
It's probably because the WebView is not updated. You should check on real devices.

**Reply by Myst3XFS** [reply]
ok, I'll do it!

**Reply by KingNothing81** [reply]
What a nice surprise. =)
Unfortunately I won't have time to test it today. But I will tomorrow evening. And then report back.

**Reply by Myst3XFS** [reply]
Tested, it worked very well, when I grow up I want to be like you.

**Reply by Myst3XFS** [reply]
Do you think it's possible and likely to use Geckoview when implementing a TV application? It might be a good alternative, considering that some TV systems are very outdated when it comes to webview.

PS: I've never used Geckoview before, I'm a novice in the subject, just questioning.

https://mozilla.github.io/geckoview/

**Reply by KingNothing81** [reply]
@Myst3XFS, @redphx:  I've also thought about if it would be possible to make the app use other render engines. That other Android app out there has an option to change render engine in the setttings. So you can switch between "System WebView", "GeckoView" and "Chrome" (if installed).

Google's Chrome browser will be (pre-) installed on many Android devices anyway. So if it would be possible to make the BxC app also use Chrome's browser engine, this could solve the issues some have when using the system's WebView. But this would still have the advantage that you don't have to install an additional browser like Kiwi because of extension support.

I also found this:
Try out WebView Beta, Dev, or Canary
https://chromium.googlesource.com/chromium/src/+/master/android_webview/docs/prerelease.md

**Reply by Myst3XFS** [reply]
I tested geckoview, xcloud ran perfectly on several devices, fast and with quality, however I was still unable to inject the script due to the different format for applying it.

PS: the size of the app was relatively small and ran independently of the devices' system resources

https://firefox-source-docs.mozilla.org/mobile/android/geckoview/consumer/geckoview-quick-start.html
https://firefox-source-docs.mozilla.org/mobile/android/geckoview/consumer/web-extensions.html

**Reply by KingNothing81** [reply]
@Myst3XFS: I don’t know if you already know, but Firefox for Android has limited extension support. Including Tampermonkey. 

https://addons.mozilla.org/de/android/addon/tampermonkey/

So if you want to check out “Better xCloud” with GeckoView, you could try it by installing Firefox.

**Reply by redphx** [reply]
The xCloud and Better xCloud are optimized for Chromium and Safari only. Firefox may work but it lacks some API.

The idea is to use ExoPlayer to display the streaming video instead of WebView, but it requires a lot more work.

**Reply by Myst3XFS** [reply]
I didn't know about exoplayer, I'll study it.

**Reply by KingNothing81** [reply]
@Myst3XFS: The good thing about using the System WebView is that on a more recent Android OS it's a seperate system package. So it will get updates over the PlayStore. I can see that it gets updated quite often on my Shield TV. Google did that so they can update the WebView without the need to to roll out a whole OS update.

So if you don't have a streaming device or Smart TV that runs a very outdated Android TV OS, the System Webview will get regular updates over PlayStore. And as H.264 is still a very common codec in online video streaming, I'm pretty sure that most if not all Android TV devices should support hardware accelerated H.264 decoding.

But I could imagine that especially on streaming devices with a small form factor, a bad "wifi performance" could be a problem for cloud gaming. Because when using apps like Netflix, the app can just buffer video to compensate fluctuating data rates. But that's not possible with cloud gaming. So I think if possible you should always use a wired internet connection. My wifi router is just behind the wall. So I'm glad that just using wifi on my device works fine.
